proposals/0000-no-sugared-incompleteness.rst

```diff
@@ -0,0 +1,144 @@
+``NoSugaredIncompleteness``
+==============
+
+.. author:: John Ericson
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
+            **After creating the pull request, edit this file again, update the
+            number in the link, and delete this bold sentence.**
+.. contents::
+
+Provide an extension to disallow syntax that would desugar to run-time errors without a ``-fdeferred-*`` flag.
+
+Motivation
+----------
+
+Going back to Milner's "Well-Typed expressions do not go wrong", typed functional programmers, and Haskellers in particular, usually prefer their programs to be free of run-time failures stemming from programmer error.
+For those learning Haskell, learning this mindset of being fastidious and aiming for code that is correct-by-construction about even banal errors is sometimes as important and novel as learning the language itself.
+We should help those learners by trying to guide them in that direction every step of the code they write.
+
+Unfortunately, the compiler and language do always help with this.
+We have old syntax that can not only introduce run-time errors, but does so implicitly.
+Incomplete pattern matches, incomplete record updates, etc., are all seldom-asked-for "conveniences" that just get in the way of a programmer trying to understand where their program can "go wrong".
+
+I think most of us would agree with the above paragraph.
+But many of us also would say it's a solved problem, because we have ``-Werror=...`` to catch all these things.
+Why write yet another proposal, for yet another language extension, when this is solved problem?
+
+Well, I just don't think ``-Werror=...`` is the right mechanism to rely on.
+This is, of course, a matter of opinion, but to me warnings are for "extra" analyses, not "essential" ones like type checking which enforce the language proper.
+Deciding whether e.g. pattern matching is exhaustive feels to me like one of those essential analyses.
+I am asking that we switch our vantage point and instead of considering programs with implicit partiality well-formed programs that are iffy, consider them ill-formed programs that we can permissively accept anyways.
+
+So, what's the practical upshot of this lawyering?
+And what mechanism might we use rather than ``-Werror=...``?
+There is another mechanism for controlling run-time errors which I think is a perfect model for how these programs should be dealt with: ``-fdefered-type-errors``.
```

Comment:
There's no `-fdefered-type-errors`, only `-fdefer-type-errors` and `-Wdeferred-type-errors`.